src/persist-client/src/read.rs

```diff
@@ -919,18 +933,59 @@ where
     /// Grab the next batch of consolidated data.
     pub async fn next(
         &mut self,
-    ) -> Option<impl Iterator<Item = ((Result<K, String>, Result<V, String>), T, D)> + '_> {
-        let iter = self
-            .consolidator
-            .next()
-            .await
-            .expect("fetching a leased part")?;
-        let iter = iter.map(|((k, v), t, d)| {
-            let key = K::decode(k, &self.read_schemas.key);
-            let val = V::decode(v, &self.read_schemas.val);
-            ((key, val), t, d)
-        });
-        Some(iter)
+    ) -> Option<
+        Box<dyn Iterator<Item = ((Result<K, String>, Result<V, String>), T, D)> + Send + Sync + '_>,
```

Comment:
This theoretically has a performance impact, since it means iterator operations will be dynamically dispatched. So far the benchmarks don't show any regressions, though! (And IIUC this not used in a hot path.)

I think the alternative is to implement a custom iterator which wraps the two options, if that seems too risky?

Reply:
Keeping it as a `Box<dyn Iterator>` works for me! If need be though it looks like we can use [`itertools::Either`](https://docs.rs/itertools/latest/itertools/enum.Either.html)

Reply:
Ah yeah, great point! I've switched to itertools since it's already a dep here... I'm still not super worried about the overhead, but now we have to worry even less.